src/rgw/rgw_common.h

```diff
@@ -1074,6 +1074,16 @@ struct RGWBucketInfo {
 
   bool empty_sync_policy() const;
 
+  rgw::BucketLayout& get_layout() { return layout; }
+  const rgw::BucketLayout& get_layout() const { return layout; }
+  bool is_indexless() const {
+    return layout.current_index.layout.type ==
+      rgw::BucketIndexType::Indexless;
+  }
+  rgw::bucket_log_layout_generation* last_log_layout_gen_p() {
+    return layout.logs.empty() ? nullptr : &layout.logs.back();
+  }
```

Comment:
```
if (bucket_info.layout.logs.empty()) {
  return -EINVAL;
}
const auto x& = bucket_info.layout.logs.back();
```

vs.

```
auto x_p = bucket_info.last_log_layout_gen_p();
if (!p) {
  return -EIO;
}
```

If it's the length of chars, we can remove "_gen" maybe.

And again, why do ~34 code sites across multiple files need to know a) the hierarchy of structs, b) that it ends in a vector, c) that the latest entry is the last element of the vector rather than the first?

Reply:
BTW, I initially thought I'd use `std::optional` but those can't be used with references. So the preferred way is to use a pointer.